src/Functions/FunctionHelpers.cpp

```diff
@@ -120,73 +119,57 @@ void validateArgumentsImpl(const IFunction & func,
+    if (column_validator && (column == nullptr || !column_validator(*column)))
         return ErrorCodes::ILLEGAL_COLUMN;
 
     return 0;
 }
 
-void validateFunctionArgumentTypes(const IFunction & func,
-                                   const ColumnsWithTypeAndName & arguments,
-                                   const FunctionArgumentDescriptors & mandatory_args,
-                                   const FunctionArgumentDescriptors & optional_args)
+void validateFunctionArguments(const IFunction & func,
+                               const ColumnsWithTypeAndName & arguments,
+                               const FunctionArgumentDescriptors & mandatory_args,
+                               const FunctionArgumentDescriptors & optional_args)
 {
     if (arguments.size() < mandatory_args.size() || arguments.size() > mandatory_args.size() + optional_args.size())
     {
-        auto join_argument_types = [](const auto & args, const String sep = ", ")
-        {
-            String result;
-            for (const auto & a : args)
-            {
-                using A = std::decay_t<decltype(a)>;
-                if constexpr (std::is_same_v<A, FunctionArgumentDescriptor>)
-                {
-                    if (a.argument_name)
-                        result += "'" + std::string(a.argument_name) + "' : ";
-                    if (a.expected_type_description)
-                        result += a.expected_type_description;
-                }
-                else if constexpr (std::is_same_v<A, ColumnWithTypeAndName>)
-                    result += a.type->getName();
-
-                result += sep;
-            }
-
-            if (!args.empty())
-                result.erase(result.end() - sep.length(), result.end());
-
-            return result;
-        };
+        auto argument_singular_or_plural = [](const auto & args){ return fmt::format("argument{}", args.size() != 1 ? "s" : ""); };
 
         throw Exception(ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH,
-            "Incorrect number of arguments for function {} provided {}{}, expected {}{} ({}{})",
-            func.getName(), arguments.size(), (!arguments.empty() ? " (" + join_argument_types(arguments) + ")" : String{}),
-            mandatory_args.size(), (!optional_args.empty() ? " to " + std::to_string(mandatory_args.size() + optional_args.size()) : ""),
-            join_argument_types(mandatory_args), (!optional_args.empty() ? ", [" + join_argument_types(optional_args) + "]" : ""));
+            "An incorrect number of arguments was specified for function '{}'. Expected {}, got {}",
+            func.getName(),
+            (!mandatory_args.empty() && !optional_args.empty())
+                ? fmt::format("{} mandatory {} and {} optional {}", mandatory_args.size(), argument_singular_or_plural(mandatory_args), optional_args.size(), argument_singular_or_plural(optional_args))
+                : (!mandatory_args.empty() && optional_args.empty())
```

Comment:
notice how because of the previous check it's enough checking just that `!mandatory_args.empty()`